submariner/multiple-cable-drivers.md

```diff
@@ -0,0 +1,136 @@
+required.
+
+## Proposal
+
+This enhancement proposes a mechanism to allow users to specify their intent as to which clusters to connect and how to
+connect them. The default would still be a full mesh with a homogeneous cable driver.
+
+The standard mechanism in Kubernetes to specify intent is by creating a Kubernetes API resource whose effect is eventually
+realized into the cluster's desired state. To extend the Kubernetes API, one could either use the generic Kubernetes
+`ConfigMap` or use a custom resource definition (CRD). There's pros and cons to each but a CRD seems more suitable for
+this enhancement.
+
+This enhancement will propose a single API for finer-grained connectivity and multiple cable driver support.
+
+### Finer-grained connectivity API
+
+This following CRD is proposed:
+
+```Go
+type ClusterConnectionPolicies struct {
+    metav1.TypeMeta `json:",inline"`
+
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec ClusterConnectionPolicy `json:"spec"`
+}
+
+type ClusterConnectionPolicySpec struct {
+    // LeftClusterSelector identifies the Cluster resources representing the clusters on one end of a connection. An empty
+    // selector indicates wildcard, ie matches any cluster.
+    LeftClusterSelector metav1.LabelSelector `json:"leftClusterSelector,omitempty"`
+
+    // RightClusterSelector identifies the Cluster resources representing the clusters on the other end of a connection.
+    // An empty selector indicates wildcard, ie matches any cluster.
+    RightClusterSelector metav1.LabelSelector `json:"rightClusterSelector,omitempty"`
+
+    // ClusterSelector identifies the Cluster resources representing a subset of clusters. not used in combination with
+    // LeftClusterSelector or RightClusterSelector
+    // An empty selector indicates wildcard, ie matches any cluster.
+    ClusterSelector metav1.LabelSelector `json:"clusterSelector,omitempty"`
+
+    CableDriver *CableDriverSpec `json:"cableDriver,omitempty"`
+    
+    priority  *int     `json:"priority,omitempty"`
+}
+
+type CableDriverSpec struct {
+    // Name of the cable driver implementation.
+    Name string `json:"name"`
+
+    // Options specifies key/value configuration parameters for the cable driver.
+    Options map[string]string `json:"options,omitempty"`
+}
+```
+
+The `ClusterConnectionPolicies` CR defines the criteria for selecting which cluster pairs/subsets can connect and associated optional
+configuration options to use for each cluster connection. To resolve the connectivity policy for a remote cluster's Endpoint,
+Submariner would search for a policy whose left and right label selectors match the corresponding Cluster resources of the local
+and remote Endpoints. If any matching policy is found, the clusters are connected. If more than one matching policy is found, the
+priority field is used to determine which policy to use. If no matching policy is found, the default cable driver and policy (full mesh)
+are used.
```

Comment:
If I read this correctly, the policies are considered on an Endpoint pair basis and always fall back to default cable driver when no policy is found. The result will always be a full mesh I think. Do we need a "don't-connect" policy that could either be applied with selectors or as the default policy?

Reply:
Yeah - good idea. I will add something to the policy config